src/com/amazon/corretto/crypto/provider/ConcatenationKdfSpi.java

```diff
@@ -0,0 +1,122 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+class ConcatenationKdfSpi extends KdfSpi {
+  private final int digestCode;
+  // Determines if the digest algorithm should be used as backing PRF or the HMAC.
+  private final boolean digestAsPrf;
+
+  ConcatenationKdfSpi(final int digestCode, final boolean digestAsPrf) {
+    this.digestCode = digestCode;
+    this.digestAsPrf = digestAsPrf;
+  }
+
+  @Override
+  protected SecretKey engineGenerateSecret(final KeySpec keySpec) throws InvalidKeySpecException {
+    if (!(keySpec instanceof ConcatenationKdfSpec)) {
+      throw new InvalidKeySpecException("Expected a key spec of type GenericSpec.");
```

Comment:
"Expected a KeySpec of type ConcatenationKdfSpec."

Reply:
Done.